### ISSUE TITLE
[RPC]: new RPC endpoint to retrieve virtual counters

### ISSUE THREAD

**Issue by krlosMata** [0xPolygon/cdk-erigon#459]
# Rationale
- Being able to get virtual zk-counters when a transaction is executed
- Dapps or any user to get estimated zk-counters to check any possible limitations on their transactions

# Implementation
- [This implementation](https://github.com/0xPolygonHermez/zkevm-node/blob/develop/jsonrpc/endpoints_zkevm.openrpc.json#L428) has been implemented in the `zkevm-node`. It mimics ethereum RPC call `eth_estimateGas`, it just add additional information related to L2
- JSON schema could be useful as a starting point

**Comment by krlosMata** [comment]
Following on this topic:
- RPC to have another type `debug tracer` very similar to [opcode logger](https://geth.ethereum.org/docs/developers/evm-tracing/built-in-tracers#struct-opcode-logger) with additional information related to virtual zk-counters
  - add all the virtual zk-counters that has been consumed by each opcode
- This information could be used to analyze and optimize top virtual zk-counters offenders

**Comment by V-Staykov** [comment]
Here is a JSON Schema for the estimateCounters method: (https://github.com/0xPolygonHermez/cdk-erigon/blob/aeb5745797968e0a511882e3b6f8e0d54b79208b/cmd/rpcdaemon/commands/endpoints_zkevm.openrpc.json)

As for the tracer with counters - we could do it but not very accurate, because we would need SMT depth for every block, which we currently don't have.

**Comment by krlosMata** [comment]
- How do you handle the standard `opcode tracer` ? Do you re-execute the transaction and get all details or it is somehow cached when the tx is being processed ?
  - if you re-execute the tx, you need to access the SMT to access data. Therefore, counters could be very accurate
  - if ot is cached in some way, maybe add more info cached, like worst `nLevel` accessed in the SMT. And use this `nLevel` when the trace is generated

**Comment by V-Staykov** [comment]
It is reexecuted, but we don't get SMT data in there. We get state data separately from the SMT that is for that block the TX is in.

One way we can go is: save the SMT depth wherever we have it and then in the trace try to get it, and if we don't bhave it - exagerate it. Also we could have a parameter for the RPC method to input the depth and use it if it is suplied.

**Comment by krlosMata** [comment]
oh! just for curiosity...how do you get state-data without querying the SMT ?

yep...I think that probably makes sense to:
- save the deepest level accessed during tx execution `deepest SMT level`
- if value is present --> use that value
- otherwise, exaggerate it (not sure how though)

**Comment by V-Staykov** [comment]
We've added a trace with counters. It looks like this:
`"smtLevels": 257,
		"structLogs": [
			{
				"pc": 0,
				"op": "PUSH1",
				"gas": 61432,
				"gasCost": 3,
				"depth": 1,
				"stack": [],
				"memory": [],
				"counters": {
					"B": 0,
					"K": 0,
					"M": 0,
					"P": 0,
					"S": 0,
					"D": 0,
					"SHA": 0,
					"A": 0
				}
			},`
			
The counters for each opcode represent the counters state before execution of the opcode (as it is for all of the other data. I wanted to make it represent just the counters the current opcode used, but they are calculated after the debug print, so it is not easily doable.

I've also added the smtLevels print just ofr reference. If we don't have any data for the depth, we just put the max depth - 256. If we have some data from a previous block - we add ~10% to the depth.

There is also a depth setting to the tracer and if it is set - it will be used in the calculations.
